File: examples/echo-python/peer_server.py
#!/usr/bin/env python3
"""The echo protocol in Python, on the symmetric peer.

The difference from ``server.py`` is the last thing a non-Rust implementation
was missing: this one **asks the caller a question mid-request**. `ui/ask` is a
reverse request, sent by the responder, and a serial server cannot do it because
it cannot wait for a reply while producing one.

Driven by the Rust client:

    cargo run -p echo-protocol --bin echo-client -- \\
        --server python3 examples/echo-python/peer_server.py
"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2] / "sdks" / "python"))

from lanok import INVALID_PARAMS, Hello, Peer, Router, RpcError, stdio  # noqa: E402
from lanok._generated_echo import PROTOCOL, UI_ASK, VERSION, method  # noqa: E402


def echo(peer: Peer, params):
    if not isinstance(params, dict) or not isinstance(params.get("text"), str):
        raise RpcError("`text` must be a string", INVALID_PARAMS)

    for step in (1, 2, 3):
        peer.notify(method.ECHO_PROGRESS, {"step": step, "of": 3})

    # The reverse request. Gated on the caller having advertised it, so an
    # older host that cannot answer is not left hanging.
    if peer.supports(UI_ASK):
        answer = peer.request(method.UI_ASK, {"question": "shout?"})
        shout = answer.get("answer") == "yes"
    else:
        shout = bool(params.get("shout"))

    return {"text": params["text"].upper() if shout else params["text"]}


def main() -> None:
    router = Router().on_request(method.ECHO, echo).on_request(method.PING, lambda _p: None)
    peer = Peer(
        router,
        serve_handshake=Hello(PROTOCOL, VERSION, [UI_ASK]),
    ).connect(stdio())
    # Serve until the caller closes our stdin.
    peer.wait_closed()


if __name__ == "__main__":
    main()

File: examples/echo-typescript/peerServer.mjs
#!/usr/bin/env node
/**
 * The echo protocol in TypeScript's runtime, on the symmetric peer.
 *
 * The difference from `server.mjs` is the last thing a non-Rust implementation
 * was missing: this one asks the caller a question mid-request. `ui/ask` is a
 * reverse request, sent by the responder, and a serial server cannot do it
 * because it cannot wait for a reply while producing one.
 *
 *   cargo run -p echo-protocol --bin echo-client -- \
 *       --server node examples/echo-typescript/peerServer.mjs
 */

import { Peer, Router, RpcError, codes, stdio } from "../../sdks/typescript/dist/index.js";
import { PROTOCOL, VERSION, capability, method } from "../../sdks/typescript/dist/generatedEcho.js";

async function echo(params, peer) {
  if (typeof params !== "object" || params === null || typeof params.text !== "string") {
    throw new RpcError("`text` must be a string", codes.invalidParams);
  }

  for (const step of [1, 2, 3]) {
    peer.notify(method.echoProgress, { step, of: 3 });
  }

  // The reverse request. Gated on the caller having advertised it, so an older
  // host that cannot answer is not left hanging.
  let shout = Boolean(params.shout);
  if (peer.supports(capability.uiAsk)) {
    const answer = await peer.request(method.uiAsk, { question: "shout?" });
    shout = answer?.answer === "yes";
  }

  return { text: shout ? params.text.toUpperCase() : params.text };
}

const router = new Router().onRequest(method.echo, echo).onRequest(method.ping, () => null);

const peer = new Peer(router, {
  serveHandshake: {
    name: PROTOCOL,
    protocolVersion: VERSION,
    capabilities: [capability.uiAsk],
  },
}).connect(stdio());

// Serve until the caller closes our stdin.
await peer.closed();
